2020/11/seating/src/main.rs

```diff
@@ -0,0 +1,411 @@
+use std::cmp::min;
+use std::env;
+use std::fs;
+
+fn main() -> Result<(), std::io::Error> {
+    let args: Vec<_> = env::args().collect();
+    let input = fs::read_to_string(&args[1])?; 
+    let layout = Layout::parse(&input);
+    println!("There are {} occupied seats at steady state", layout.find_steady_state(false).occupied());
+
+    Ok(())
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Position {
+    Floor,
+    Empty,
+    Occupied,
+}
+
+pub struct Layout {
+    map: Vec<Vec<Position>>,
+}
+
+impl Layout {
+    pub fn parse(input: &str) -> Layout {
+        let mut map = Vec::new();
+        for y in input.lines() {
+            let mut row = Vec::new();
+            for x in y.chars() {
+                row.push(match x {
+                    '.' => Position::Floor,
+                    'L' => Position::Empty,
+                    '#' => Position::Occupied,
+                    _ => panic!(),
+                });
+            }
+            map.push(row);
+        }
+        Layout { map: map }
+    }
+
+    fn apply_round(&self, adjacent: bool) -> Layout {
+        let mut map = Vec::new();
+        for y in 0..self.map.len() {
+            let mut row = Vec::new();
+            for x in 0..self.map[y].len() {
+                let new = match adjacent {
+                    true => self.transform_adjacent(y, x),
+                    false => self.transform_visible(y, x),
+                };
+                row.push(new);
+            }
+            map.push(row);
+        }
+        Layout { map: map }
+    }
+
+    fn transform_adjacent(&self, y: usize, x: usize) -> Position {
+        let cur = self.map[y][x];
+        let adjacent_occupied = self.adjacent_occupied(y as isize, x as isize);
+        match cur {
+            Position::Empty => match adjacent_occupied {
+                0 => Position::Occupied,
+                _ => cur,
+            },
+            Position::Occupied => match adjacent_occupied >= 4 {
+                true => Position::Empty,
+                _ => cur,
+            },
+            _ => cur,
+        }
+    }
+
+    fn transform_visible(&self, y: usize, x: usize) -> Position {
+        let cur = self.map[y][x];
+        let visible_occupied = self.visible_occupied(y, x);
+        match cur {
+            Position::Empty => match visible_occupied {
+                0 => Position::Occupied,
+                _ => cur,
+            },
+            Position::Occupied => match visible_occupied >= 5 {
+                true => Position::Empty,
+                _ => cur,
+            },
+            _ => cur,
+        }
+    }
+
+    fn adjacent_occupied(&self, y: isize, x: isize) -> usize {
+        let mut occupied = 0;
+
+        for i in 0..3 {
+            occupied += self.one_if_occupied(y - 1, x + i - 1);
```

Comment:
If you wanted to get *super* zesty, you could implement `Into<usize>` for `Position` and then `self.one_if_occupied(x, y)` just becomes `self.get(x, y).into()`